README.md

```diff
@@ -216,223 +216,59 @@ interface StaticModuleRecord {
-  // tool for a REPL, such that a const or let binding from one evaluation
-  // persists to the next.
-  // Subsequent proposals might afford other modes, like sloppy globals mode.
-  // TODO is sloppyGlobalsMode only sensible in the context of the shim?
-  evaluate(source: string): any;
-
-  // load causes a loader to load module descriptors for the
-  // transitive dependencies of a specified module into its
-  // memo, but does not initialize any modules.
-  // The load function is useful for tools like bundlers and importmap
-  // generators that load a module graph in an emulated host environment
-  // but cannot and should not emulate evaluation.
-  async load(fullSpecifier: string): Promise<void>
-
-  // import induces the asynchronous load phase and then initializes
-  // the given module and any of its transitive dependencies that
-  // have not already begun initialization.
-  async import(fullSpecifier: string): Promise<ModuleExportsNamespace>;
-
-  // TC53: some embedded systems hosts exclude promises and so require these
-  // synchronous variants of import and load.
-  // On hosts where these functions cannot succeed synchronously,
-  // they must throw an error without the side effect of initializing any
-  // additional modules.
-  // For modules that have a top-level await, these would return
-  // after the module first awaits but not after any subsequent promise queue
-  // job.
-
-  // If a host supports both importNow and import, they must share
-  // a memo of full specifiers to promises for the module exports namespace.
-  // importNow and loadNow followed by import or load must not accidentally
-  // reinitialize the underlying module or produce different namespaces.
-
-  importNow(fullSpecifier: string): ModuleExportsNamespace;
-
-  loadNow(fullSpecifier: string): void;
+// ModuleInstance reifies an entangled pair of module environment record
+// and module exports namespace from a particular array of bindings
+// that correspond to the `import` and `export` declarations of a
+// module.
+interface ModuleInstance {
+  // Creates a module instance for either a static module record
+  // or a synthetic static module record.
+  // Creates a module environment record that defers to the
+  // designated global environment record or that associated
+  // with the ModuleInstance constructor.
+  // The module instance will delegate to the given dynamic import
+  // and will use the given global for all eval and Function constructor
+  // behaviors.
+  constructor(
+    record: StaticModuleRecord | SyntheticStaticModuleRecord,
+    {
+      import?: (importSpecifier: string) => Promise<ModuleExportsNamespace>,
+      importMeta?: Object,
+      global?: Global,
+    }
+  );
+
+  // Links the module instance with another module instance
+  // for one of the importSpecifiers declared in the constructed
+  // bindings.
+  link(importSpecifier: string, instance: ModuleInstance);
```

Comment:
How can this new design handles dynamic import?

Reply:
A dynamic import hook must be passed as a ModuleInstance constructor option, which will throw if the static module record has `needsImport` set and `import` is not provided.

Reply:
This design seems right to me - in that the `HostImportModuleDynamically` hook effectively needs to be exposed as a function for each individual module instance in this way (along with the globals). Nice to see we agree on how it fleshes out!

Also agreed, the repeating of this information for all modules in the same compartment is clearly simplified by having a singular loader, which in turn removes the need for the dynamic import hook even being necessary as well - and these are strong arguments for a loader over an unmanaged instance model.

Reply:
Why would you want to be able to get different results between dynamically and statically importing the same specifier? (modulo thenable modules)

Reply:
You certainly wouldn't, which is also another possible argument for loader-level over instance-level management.

Reply:
@cardiy is also a proponent of exposing `ModuleInstance` and not exposing a `Loader`. A difference he proposes is that `ModuleInstance` would use an `importHook(importSpecifier, moduleInstance) => Promise<ModuleInstance>` hook for both static and dynamic import. There would need to be a `link()` method with no arguments that drives the `importHook` for static imports before initialization.

Caridy believes it would be sufficient for the `ModuleInstance` implementation to enforce local consistency of static and dynamic import by memoizing `importHook` locally. I also find that sufficient, but because it would be a more significant memory burden, and because the `importHook` would presumably want to enforce stronger consistency invariants among modules, I would be open to deferring the responsibility of maintaining the consistency invariant to the user-provided hook.

For the purposes of isolating guest code, the `ModuleInstance` would likewise need to accept an optional alternate global environment and I have not reckoned out whether that would be sufficient to implement a compartmental loader in user code.

Please let us know what you think. I will probably sketch that alternative view for purposes of discussion. I’m attracted to the idea of not having to convince TC39 of the exact shape of module descriptors, if that’s a possibility, though the `Compartment` API is still quite convenient.